### ISSUE TITLE
Lab2 DDP

### ISSUE THREAD

**Issue by shwgao** [shwgao/CS-ECE-599-labs#3]
Task 1:
The output is a bit messy. Please tidy it up as follows:
[GPU0] Epoch 1 | Batch size: 64 | Steps: (1/39) | Loss: 0.032 | Accuracy: 12%
[GPU0] Epoch 1 | Batch size: 64 | Steps: (5/39) | Loss: 0.019 | Accuracy: 12%
…

Please leave comments on the GitHub issue and include the last four digits of your ONID. To make extra credits, open a pull request of your solution for these tasks:

Task 2:
Predict the three samples in the directory ‘lab2’?

Task 3:
Test the model with test data in the dataset. What are some strategies to improve the accuracy of this application?

Task 4:
Does the DDP strategy accelerate the training process? If not, why not?

**Comment by shwgao** [comment]
[GPU0] Epoch 0 | Batchsize: 64 | Steps:0/39 | Loss: 5.3000 | Accuracy: 0.0000
[GPU0] Epoch 0 | Batchsize: 64 | Steps:10/39 | Loss: 5.2686 | Accuracy: 0.0000
[GPU0] Epoch 0 | Batchsize: 64 | Steps:20/39 | Loss: 4.6233 | Accuracy: 0.0000
[GPU0] Epoch 0 | Batchsize: 64 | Steps:30/39 | Loss: 4.5893 | Accuracy: 0.0000
Epoch 0 took 23.70 seconds
Epoch 0 | Training checkpoint saved at checkpoint.pt
[GPU0] Epoch 1 | Batchsize: 64 | Steps:0/39 | Loss: 37.8242 | Accuracy: 2.2177
[GPU0] Epoch 1 | Batchsize: 64 | Steps:10/39 | Loss: 4.8106 | Accuracy: 2.2177
[GPU0] Epoch 1 | Batchsize: 64 | Steps:20/39 | Loss: 4.7749 | Accuracy: 2.2177

**Comment by pavlohav** [comment]
Without pretraining. Loss was still very high

<img width="590" alt="Image" src="https://github.com/user-attachments/assets/b3f71c7a-a3cc-487a-9e6c-a1e891bb4cf1" />

With pretraining. Loss began to drop off very quickly. Training began to take a very long time

<img width="564" alt="Image" src="https://github.com/user-attachments/assets/c37774e1-965f-4a9a-a847-d4e129efec1a" />

Best loss i was able to get was 0.0723 and an accuracy of 96.6324 before my session expired

OSU id: 7425

**Comment by megant2321** [comment]
There are multiple strategies to improve accuracy of model. More data is an obvious answer. You can create more data in a number of ways, for example, in our term project we are creating more data by creating data sets with overlapping data. This almost doubles the amount of data we have. The DDP training will accelerate the training process especially with a wider dataset due to the parallelism. But if the dataset is kept small the parallelism will actually increase the training process time as it requires more gpus.
OSU ID:7072

**Comment by drcrockerjr** [comment]
One strategy to improve accuracy is to add more data preconditioning. This entails creating transformations of the data that could give the model a more holistic view on the problem being train for. The DDP strategy allow for 2 seperate models to preform training on separate GPUs. With this parallelization, the training time for a whole data set can be split in half.

OSU ID: 3645

**Comment by NotJeremyCapps** [comment]
To improve accuracy we can add more data for our model to train on, another way is to create data overlap, creating more training points with the same amount of data. To help accommodate for the increased amount of training points we can use DPP to utilize multiple GPUs in our training process to accelerate training.
ONID:1556

**Comment by Maxttxxtt0110** [comment]
To improve the model accuracy, we should put more data for the training or we can pre-initializing the image transformation pipeline for recreating it every time. DDP did improve the efficiency while splitting workload into multiple GPUs

ONID:5702
